File: src/main/java/dev/thornium/flatcore/utils/FTIconSetUtils.java
package dev.thornium.flatcore.utils;

import com.gregtechceu.gtceu.GTCEu;
import com.gregtechceu.gtceu.api.data.chemical.material.Material;
import com.gregtechceu.gtceu.api.data.chemical.material.info.MaterialIconSet;
import dev.thornium.flatcore.registry.FTMaterialIconSet;
import net.minecraft.resources.ResourceLocation;

import java.util.Optional;

public final class FTIconSetUtils {
    public static String overridePipePath(String original, Material material) {
        return getOverrideIconSet(material)
                .map(iconSet -> replacePath(original, iconSet))
                .orElse(original);
    }

    public static ResourceLocation overridePipeLocation(ResourceLocation original, Material material) {
        return getOverrideIconSet(material)
                .map(iconSet -> GTCEu.id(replacePath(original.getPath(), iconSet)))
                .orElse(original);
    }

    public static Optional<MaterialIconSet> getOverrideIconSet(Material material) {
        MaterialIconSet iconSet = material.getMaterialIconSet();
        return FTMaterialIconSet.ALL.contains(iconSet) ? Optional.of(iconSet) : Optional.empty();
    }

    private static String replacePath(String original, MaterialIconSet iconSet) {
        return original.replace("block/pipe", "block/material_sets/" + iconSet.name);
    }
}

File: src/main/kotlin/dev/thornium/flatcore/utils/FTIconSetUtils.kt
package dev.thornium.flatcore.utils

import com.gregtechceu.gtceu.GTCEu
import com.gregtechceu.gtceu.api.data.chemical.material.Material
import com.gregtechceu.gtceu.api.data.chemical.material.info.MaterialIconSet
import dev.thornium.flatcore.registry.FTMaterialIconSet
import net.minecraft.resources.ResourceLocation

object FTIconSetUtils {
    @JvmStatic
    fun overridePipePath(original: String, material: Material) =
        getOverrideIconSet(material)
            ?.let { iconSet -> replacePath(original, iconSet) }
            ?: original

    @JvmStatic
    fun overridePipeLocation(original: ResourceLocation, material: Material) =
        getOverrideIconSet(material)
            ?.let { iconSet -> GTCEu.id(replacePath(original.path, iconSet)) }
            ?: original

    @JvmStatic
    fun getOverrideIconSet(material: Material) =
        FTMaterialIconSet.ALL.firstOrNull { it.name == material.materialIconSet.name }

    private fun replacePath(original: String, iconSet: MaterialIconSet) =
        original.replace("block/pipe", "block/material_sets/${iconSet.name}")
}
